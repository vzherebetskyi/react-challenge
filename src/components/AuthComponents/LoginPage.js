import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { initializeApp } from 'firebase/app';
import {
  createUserWithEmailAndPassword,
  getAuth,
  updateProfile,
} from 'firebase/auth';

import AuthComponentWrapper from './AuthComponentWrapper';
import { handleChangeEmail, startLoginUser } from '../../actions/auth';
import { handleShowNotification } from '../../actions/notifications';
import { errorMessages } from '../../utils/constants';
import { useEnterListener } from '../../customHooks/useBtnListener';
import { generateUser } from '../../utils/utilFunctions';
import Mail from '../../assets/images/mail.svg';
import PasswordInput from '../reusedComponents/PasswordInput';

const LoginPage = props => {
  const dispatch = useDispatch();
  const userEmail = useSelector(state => state.auth.userEmail);
  const enterPressed = useEnterListener();

  const [credentials, setCredentials] = useState({
    email: userEmail,
    password: '',
  });
  const [errors, setErrors] = useState({
    email: '',
    password: '',
  });
  const [useFirebase, setUseFirebase] = useState(false);

  useEffect(() => {
    if (enterPressed) {
      handleValidate();
    }
  }, [enterPressed]);

  const handleChangeInput = val => {
    const newCredentials = { ...credentials };
    const name = val.name;
    newCredentials[name] = val.value;
    setCredentials(newCredentials);
    if (name === 'email') {
      dispatch(handleChangeEmail(val.value));
    }
    if (errors[name]) {
      const newErrors = { ...errors, [name]: '' };
      setErrors(newErrors);
    }
  };

  const handleValidate = () => {
    let errorIsPresent;
    const newErrors = { ...errors };
    for (const el in credentials) {
      if (credentials[el].length === 0) {
        errorIsPresent = 1;
        newErrors[el] = errorMessages.notEmptyField;
      }
    }
    setErrors(newErrors);
    if (!errorIsPresent) {
      if (useFirebase) {
        dispatch(startLoginUser(credentials.email, credentials.password))();
      } else
        dispatch(
          handleShowNotification('Provided email and password is/are incorrect')
        );
    }
  };

  const switchToSignUp = () => {
    props.history.push('/register');
  };

  const handleGenerateUser = async () => {
    const generatedData = await generateUser();

    if (generatedData && generatedData.results) {
      const userName = generatedData.results[0].login.username;
      const email = generatedData.results[0].email;
      const password =
        generatedData.results[0].login.password +
        generatedData.results[0].login.username;
      const config = {
        apiKey: process.env.FIREBASE_API_KEY,
        authDomain: process.env.FIREBASE_AUTH_DOMAIN,
        databaseURL: process.env.FIREBASE_DATABASE_URL,
      };
      const secondaryApp = initializeApp(config, 'Secondary');
      const auth = getAuth(secondaryApp);
      createUserWithEmailAndPassword(auth, email, password)
        .then(userCredential => {
          console.log('success', userCredential);
          dispatch(handleShowNotification('Signed up successfully'));

          return updateProfile(auth.currentUser, {
            displayName: userName,
          });
        })
        .then(() => {
          console.log('User profile successfully updated');
          const newCredentials = {
            email: email,
            password: password,
          };
          setCredentials(newCredentials);
        })
        .catch(error => {
          console.log(error.message);
          dispatch(handleShowNotification(error.message));
        });
    }
  };

  return (
    <AuthComponentWrapper title="LOGIN" history={props.history}>
      <form
        onKeyPress={e => e.key === 'Enter' && e.preventDefault()}
        onSubmit={e => {
          e.preventDefault();
          handleValidate();
        }}
      >
        <div className="input-container">
          <img src={Mail} alt="mail" />
          <input
            autoComplete="off"
            placeholder="E-Mail Address"
            type="text"
            name="email"
            value={credentials.email}
            onChange={e =>
              handleChangeInput({ name: e.target.name, value: e.target.value })
            }
          />
          {errors.email && <p className="error-message">{errors.email}</p>}
        </div>
        <PasswordInput
          handleChangeInput={handleChangeInput}
          credentials={credentials}
          errors={errors}
          inputName="password"
          inputPlaceholder="Password"
        />
        <div className="checkbox-container">
          <input
            type="checkbox"
            id="useFirebase"
            name="useFirebase"
            checked={useFirebase}
            onChange={() => setUseFirebase(prevState => !prevState)}
          />
          <label htmlFor="useFirebase">
            I would like to use Firebase for login
          </label>
        </div>
        <div className="parall-styled-btn-container">
          <div onClick={handleValidate} className="parall-styled-btn">
            <button type="submit">LOGIN</button>
          </div>
        </div>
      </form>
      <div className="auth-container-bottom-text">
        <p>
          Don't have an account yet?{' '}
          <span onClick={switchToSignUp}>Signup</span>
        </p>
        <p>Don't remember your password? Recover my Password</p>
        <p>
          <span
            style={{ color: '#F8D126' }}
            onClick={!useFirebase ? null : handleGenerateUser}
          >
            Generate a custom user.
          </span>{' '}
          Works with Firebase. See FAQ for more details
        </p>
      </div>
    </AuthComponentWrapper>
  );
};

export default LoginPage;
