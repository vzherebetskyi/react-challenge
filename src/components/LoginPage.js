import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { initializeApp } from 'firebase/app';
import {
  createUserWithEmailAndPassword,
  getAuth,
  updateProfile,
} from 'firebase/auth';

import { handleChangeEmail, startLoginUser } from '../actions/auth';
import { handleShowNotification } from '../actions/notifications';
import { errorMessages } from '../utils/constants';
import { useEnterListener } from '../customHooks/useBtnListener';
import { generateUser } from '../utils/utilFunctions';

const LoginPage = props => {
  const dispatch = useDispatch();
  const userEmail = useSelector(state => state.auth.userEmail);
  const enterPressed = useEnterListener();

  const [credentials, setCredentials] = useState({
    email: userEmail,
    password: '',
  });
  const [errors, setErrors] = useState({
    email: '',
    password: '',
  });
  const [useFirebase, setUseFirebase] = useState(false);

  useEffect(() => {
    if (enterPressed) {
      handleValidate();
    }
  }, [enterPressed]);

  const handleChangeInput = val => {
    const newCredentials = { ...credentials };
    const name = val.name;
    newCredentials[name] = val.value;
    setCredentials(newCredentials);
    if (name === 'email') {
      dispatch(handleChangeEmail(val.value));
    }
    if (errors[name]) {
      const newErrors = { ...errors, [name]: '' };
      setErrors(newErrors);
    }
  };

  const handleValidate = () => {
    let errorIsPresent;
    const newErrors = { ...errors };
    for (const el in credentials) {
      if (credentials[el].length === 0) {
        errorIsPresent = 1;
        newErrors[el] = errorMessages.notEmptyField;
      }
    }
    setErrors(newErrors);
    if (!errorIsPresent) {
      if (useFirebase) {
        dispatch(startLoginUser(credentials.email, credentials.password))();
      } else
        dispatch(
          handleShowNotification('Provided email and password is/are incorrect')
        );
    }
  };

  const switchToSignUp = () => {
    props.history.push('/register');
  };

  const handleGenerateUser = async () => {
    const generatedData = await generateUser();

    if (generatedData && generatedData.results) {
      const userName = generatedData.results[0].login.username;
      const email = generatedData.results[0].email;
      const password =
        generatedData.results[0].login.password +
        generatedData.results[0].login.username;
      const config = {
        apiKey: process.env.FIREBASE_API_KEY,
        authDomain: process.env.FIREBASE_AUTH_DOMAIN,
        databaseURL: process.env.FIREBASE_DATABASE_URL,
      };
      const secondaryApp = initializeApp(config, 'Secondary');
      const auth = getAuth(secondaryApp);
      createUserWithEmailAndPassword(auth, email, password)
        .then(userCredential => {
          console.log('success', userCredential);
          dispatch(handleShowNotification('Signed up successfully'));

          return updateProfile(auth.currentUser, {
            displayName: userName,
          });
        })
        .then(() => {
          console.log('User profile successfully updated');
          const newCredentials = {
            email: email,
            password: password,
          };
          setCredentials(newCredentials);
        })
        .catch(error => {
          console.log(error.message);
          dispatch(handleShowNotification(error.message));
        });
    }
  };

  return (
    <form
      onKeyPress={e => e.key === 'Enter' && e.preventDefault()}
      onSubmit={e => {
        e.preventDefault();
        handleValidate();
      }}
    >
      <input
        autoComplete="off"
        placeholder="E-Mail Address"
        type="text"
        name="email"
        value={credentials.email}
        onChange={e =>
          handleChangeInput({ name: e.target.name, value: e.target.value })
        }
      />
      {errors.email && <p>{errors.email}</p>}
      <input
        autoComplete="off"
        placeholder="Password"
        type="text"
        name="password"
        value={credentials.password}
        onChange={e =>
          handleChangeInput({ name: e.target.name, value: e.target.value })
        }
      />
      {errors.password && <p>{errors.password}</p>}
      <div>
        <input
          type="checkbox"
          id="useFirebase"
          name="useFirebase"
          checked={useFirebase}
          onChange={() => setUseFirebase(prevState => !prevState)}
        />
        <label htmlFor="useFirebase">
          I would like to use Firebase for login
        </label>
      </div>
      <button type="submit">Login</button>
      <button type="button" onClick={switchToSignUp}>
        SignUp
      </button>
      <button
        type="button"
        onClick={handleGenerateUser}
        disabled={!useFirebase}
      >
        Generate a user
      </button>
    </form>
  );
};

export default LoginPage;
