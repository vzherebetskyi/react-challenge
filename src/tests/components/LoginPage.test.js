import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import LoginPage from '../../components/AuthComponents/LoginPage';

afterEach(() => {
  cleanup();
  jest.resetAllMocks();
});

jest.mock('firebase/app', () => {
  return {
    initializeApp: jest.fn(),
  };
});

jest.mock('firebase/auth', () => {
  return {
    getAuth: jest.fn(),
  };
});

const mockDispatch = jest.fn(() => jest.fn());
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

const history = { push: jest.fn() };

test('should render LoginPage correctly', () => {
  const elem = renderer.create(<LoginPage history={history} />).toJSON();
  expect(elem).toMatchSnapshot();
});

test('login button click should not trigger dispatch and errors should appear', () => {
  const { getAllByText } = render(<LoginPage history={history} />);
  const loginBtn = screen.getByTestId('login-btn');
  fireEvent.click(loginBtn);
  expect(mockDispatch).toHaveBeenCalledTimes(0);
  expect(getAllByText('This field should not be empty'));
});

// test('login button click should trigger dispatch and login via firebase (second dispatch)', () => {
//   const { container } = render(<LoginPage history={history} />);
//   const loginBtn = screen.getByTestId('login-btn');
//   const checkboxFirebase = screen.getByTestId('usefirebase-checkbox');
//   fireEvent.click(checkboxFirebase);
//   const inputs = container.querySelectorAll('input');
//   fireEvent.change(inputs[0], { target: { value: 'testtest@test.com' } });
//   fireEvent.change(inputs[1], { target: { value: 'Test!1234' } });
//   fireEvent.click(loginBtn);
//   // first dispatch call for changing email, second dispatch call for login via firebase
//   expect(mockDispatch).toHaveBeenCalledTimes(2);
// });
