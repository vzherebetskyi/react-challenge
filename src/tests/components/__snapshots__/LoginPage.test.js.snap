// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LoginPage correctly 1`] = `
<div
  className="auth-container"
>
  <div
    className="auth-title-panel"
  >
    <img
      alt="main-logo"
      src={Object {}}
    />
  </div>
  <div
    className="auth-faq-button"
    onClick={[Function]}
  >
    <img
      alt="question"
      src={Object {}}
    />
  </div>
  <div
    className="form-container"
  >
    <h1
      className="base-title"
    >
      LOGIN
    </h1>
    <form
      onKeyPress={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="input-container"
      >
        <img
          alt="email"
          src={Object {}}
        />
        <input
          autoComplete="off"
          data-testid="base-input"
          name="email"
          onChange={[Function]}
          placeholder="E-Mail Address"
          type="text"
          value=""
        />
        
      </div>
      <div
        className="input-container"
      >
        <img
          alt="lock"
          src={Object {}}
        />
        <img
          alt="eye"
          className="input-container__eye"
          onClick={[Function]}
          src={Object {}}
        />
        <input
          autoComplete="off"
          data-testid="password-input"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
        
      </div>
      <div
        className="checkbox-container"
      >
        <input
          checked={false}
          id="useFirebase"
          name="useFirebase"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          data-testid="usefirebase-checkbox"
          htmlFor="useFirebase"
        >
          I would like to use Firebase for login
        </label>
      </div>
      <div
        className="parall-styled-btn-container"
      >
        <button
          className="parall-styled-btn"
          data-testid="login-btn"
          type="submit"
        >
          <div>
            LOGIN
          </div>
        </button>
      </div>
    </form>
    <div
      className="auth-container-bottom-text"
    >
      <p>
        Don't have an account yet?
         
        <span
          onClick={[Function]}
        >
          Signup
        </span>
      </p>
      <p>
        Don't remember your password? Recover my Password
      </p>
      <p>
        <span
          onClick={null}
          style={
            Object {
              "color": "#F8D126",
            }
          }
        >
          Generate a custom user.
        </span>
         
        Works with Firebase. See FAQ for more details
      </p>
    </div>
  </div>
</div>
`;
